File: guests/rust/richlabel.rs
//! The rich label scene (tools/scenes/richlabel.steps; docs/rich-text-plan.md
//! R8, §15): a label carries the inline vocabulary read-only — the app writes
//! its document and edits it, and the widget draws the runs over the role's
//! own font. THE OFFSETS ARE UTF-8 BYTES (docs/ranges-units.md).

const DOC: &str = "Héllo world, code";

#[derive(Clone)]
enum Msg {
    Seed,
    Insert,
    Mark,
}

fn spell(runs: &[kaya::Run]) -> String {
    runs.iter()
        .map(|run| {
            if run.value == "true" {
                format!("{}:{} {}", run.start, run.end, run.name)
            } else {
                format!("{}:{} {}={}", run.start, run.end, run.name, run.value)
            }
        })
        .collect::<Vec<_>>()
        .join("|")
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let (runs, body, heading) = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("richlabel");
        let runs = tx.signal("");
        let (root, (body, heading)) = tx
            .column(|tx| {
                let body_text = tx.signal("");
                let heading_text = tx.signal("Heading with italic");
                let body = tx.label(body_text).rich().a11y_id("body").id(); // label#0
                let heading = tx
                    .label(heading_text)
                    .role(kaya::Role::Heading)
                    .rich()
                    .a11y_id("heading")
                    .id(); // label#1
                tx.label(runs).a11y_id("runs"); // label#2
                tx.row(|tx| {
                    for (title, msg) in [("seed", Msg::Seed), ("insert", Msg::Insert), ("mark", Msg::Mark)] {
                        let button = tx.button(title).id();
                        msgs.on_click(button, msg);
                    }
                });
                (body, heading)
            })
            .into_parts();
        tx.mount(root);
        (runs, body, heading)
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Seed => {
                let doc = kaya::Document::new(DOC)
                    .bold(0..6)
                    .link(7..12, "https://kaya.dev")
                    .mark(14..18, "code", "true");
                let title = kaya::Document::new("Heading with italic").mark(13..19, "italic", "true");
                ctx.apply(|tx| {
                    tx.set_document(body, &doc);
                    tx.set_document(heading, &title);
                    tx.write(runs, spell(&doc.runs));
                });
            }
            Msg::Insert => {
                let edit = kaya::Edit::insert(6, ", big").mark(2..5, "italic", "true");
                ctx.apply(|tx| tx.apply_edit(body, &edit));
                let mirror = spell(&ctx.document(body).runs);
                ctx.apply(|tx| tx.write(runs, mirror.clone()));
            }
            // THE RANGED ACT ON A LABEL (docs/rich-text-plan.md §17): an italic
            // over a range and the bold taken off another, the label's own
            // document written by range; the runs label is the binding's fold.
            Msg::Mark => {
                ctx.apply(|tx| {
                    tx.format_range(body, 1..4, "italic", "true");
                    tx.unformat_range(body, 0..3, "bold"); // "Hé": byte 2 is inside the é
                });
                let mirror = spell(&ctx.document(body).runs);
                ctx.apply(|tx| tx.write(runs, mirror.clone()));
            }
        }
    }
}

fn main() {
    kaya::run(app)
}

File: guests/swift/richlabel.swift
// The rich label scene, Swift port — guests/rust/richlabel.rs,
// tools/scenes/richlabel.steps (docs/rich-text-plan.md R8, §15): a label
// carries the inline vocabulary read-only. THE OFFSETS ARE UTF-8 BYTES: the
// é in the first word is what makes a UTF-16 reader fail
// (docs/ranges-units.md).

import Foundation

let document = "Héllo world, code"
let title = "Heading with italic"

/// The core's spelling of runs (`expect_runs`), so the binding's document
/// and the core's mirror are compared as one string.
func spell(_ runs: [KayaRun]) -> String {
    runs.map { run in
        run.isFlag
            ? "\(run.range.lowerBound):\(run.range.upperBound) \(run.name)"
            : "\(run.range.lowerBound):\(run.range.upperBound) \(run.name)=\(run.value)"
    }.joined(separator: "|")
}

let app = KayaApp()

app.build { tx in
    tx.window(title: "richlabel")
    let runs = tx.signal(.str(""))

    // A widget parents at CREATION, so the two labels ride out through
    // these vars (docs/traps.md, result builders).
    var body: KayaWidget! = nil
    var heading: KayaWidget! = nil

    let root = tx.column {
        let bodyText = tx.signal(.str(""))
        let headingText = tx.signal(.str(title))
        body = tx.label(bind: bodyText, rich: true)  // label#0
        tx.setA11yId(body, "body")
        heading = tx.label(bind: headingText, role: .heading, rich: true)  // label#1
        tx.setA11yId(heading, "heading")
        let mirror = tx.label(bind: runs)  // label#2
        tx.setA11yId(mirror, "runs")
        tx.row {
            tx.button("seed") { t in  // button#0
                let doc = KayaDocument(document)
                    .bold(0..<6)
                    .link(7..<12, "https://kaya.dev")
                    .mark(14..<18, "code", true)
                let heads = KayaDocument(title).mark(13..<19, "italic", true)
                t.setDocument(body, doc)
                t.setDocument(heading, heads)
                t.write(runs, .str(spell(doc.runs)))
            }
            tx.button("insert") { t in  // button#1
                let edit = KayaEdit.insert(at: 6, ", big").mark(2..<5, "italic", true)
                t.applyEdit(body, edit)
                t.write(runs, .str(spell(app.document(body).runs)))
            }
            // THE RANGED ACT ON A LABEL (docs/rich-text-plan.md §17): the
            // label's own document written by range, no selection to move.
            tx.button("mark") { t in  // button#2
                t.formatRange(body, 1..<4, "italic", true)
                t.unformatRange(body, 0..<3, "bold")  // "Hé": byte 2 is inside the é
                t.write(runs, .str(spell(app.document(body).runs)))
            }
        }
    }
    tx.mount(root)
}

app.run()
